dvc/repo/ls.py

```diff
@@ -1,79 +1,127 @@
 
 from dvc.exceptions import PathMissingError, OutputNotFoundError
 
 
 @staticmethod
-def ls(url, target=None, rev=None, recursive=None, outs_only=False):
+def ls(
+    url, target=None, rev=None, recursive=None, outs_only=False,
+):
+    """Methods for getting files and outputs for the repo.
+
+    Args:
+        url (str): The repo url
+        target (str, optional)
+        rev (str, optional)
+        recursive (bool, optional)
+        outs_only (bool, optional): Shows only DVC-artifacts
+
+    Returns:
+        list of `entry`
+
+    Notes:
+        `entry` is a dictionary with structure
+        {
+            "path": str,
+            "isout": bool,
+            "isdir": bool,
+            "isexec": bool,
+        }
+    """
     from dvc.external_repo import external_repo
     from dvc.repo import Repo
     from dvc.utils import relpath
 
     with external_repo(url, rev) as repo:
         target_path_info = _get_target_path_info(repo, target)
-        result = []
+        fs_nodes = []
         if isinstance(repo, Repo):
-            result.extend(_ls_outs_repo(repo, target_path_info, recursive))
+            fs_nodes.extend(_ls_outs_repo(repo, target_path_info, recursive))
 
         if not outs_only:
-            result.extend(_ls_files_repo(target_path_info, recursive))
+            fs_nodes.extend(_ls_files_repo(target_path_info, recursive))
 
-        if target and not result:
+        if target and not fs_nodes:
             raise PathMissingError(target, repo, output_only=outs_only)
 
-        def prettify(path_info):
-            if path_info == target_path_info:
-                return path_info.name
-            return relpath(path_info, target_path_info)
+        fs_nodes = {n["path_info"]: n for n in fs_nodes}.values()
 
-        result = list(set(map(prettify, result)))
-        result.sort()
-    return result
+        def get_entry(fs_node):
+            path_info = fs_node["path_info"]
+            path = (
+                path_info.name
+                if path_info == target_path_info
+                else relpath(path_info, target_path_info)
+            )
+            return {
+                "path": path,
+                "isout": fs_node.get("isout", False),
+                "isdir": fs_node.get("isdir", False),
+                "isexec": fs_node.get("isexec", False),
+            }
 
+        entries = [get_entry(fs_node) for fs_node in fs_nodes]
+        entries.sort(key=lambda f: f["path"])
+    return entries
 
-def _ls_files_repo(target_path_info, recursive=None):
+
+def _ls_files_repo(path_info, recursive=None):
     from dvc.compat import fspath
     from dvc.ignore import CleanTree
     from dvc.path_info import PathInfo
     from dvc.scm.tree import WorkingTree
 
-    if not os.path.exists(fspath(target_path_info)):
+    if not os.path.exists(fspath(path_info)):
         return []
 
     files = []
-    tree = CleanTree(WorkingTree(target_path_info))
+    tree = CleanTree(WorkingTree(path_info))
```

Comment:
This will not read `.dvcignore` files from repo root to `path_info`, should simply use `repo.tree` here.

Reply:
I didn't get your point, could you elaborate it more, please?
Here `path_info` is `repo.root_dir` + `path`

Reply:
If you have a structure:

```bash
.dvc/
.dvcignore
data/
    a
    b
    c
```

And `.dvcignore` contains `c`, then your code will list all of the files in `data`, while `c` should be skipped. The reason is that `.dvcignore` file won't be read because it is outside of your scope. If you use `repo.tree` it will work correctly, also no need to use internals like `CleanTree/WorkingTree`.

Reply:
For me it's not clear from requirements that `c` should be skipped. Could you please raise a bug, cause it goes out the scope of the PR? @Suor

Reply:
If you have `.dvcignore` file inside path info you are listing then it will be read and applied if `.dvcignore` file is in dvc repo somewhere closer to the root than `path_info` then it is not read. This is inconsistent, a bug in this particular PR, not out of scope.

Reply:
You can put different way, `CleanTree` should be instantiated from repo root, doing it for any subdir is an error, which causes a bug above.

Reply:
Other thing is that `CleanTree` as a very internal one, now only `Repo` knows about it, any code outside that should not know about that to limit complexity, that was one of the ideas behind `CleanTree`. We use it in a couple of unit tests, but that's it.

Reply:
> This is inconsistent, a bug in this particular PR, not out of scope.

Current PR relates colorizing output, I've create the ticket https://github.com/iterative/dvc/issues/3431 with the behavior you described

Reply:
Looked at the current repo code and you are right, this bug was there before. Ok solve it separately.